File: jar-common/src/main/java/com/beancounter/common/contracts/FxRequest.java
package com.beancounter.common.contracts;

import com.beancounter.common.model.IsoCurrencyPair;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import java.util.ArrayList;
import java.util.Collection;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class FxRequest {
  @JsonInclude(JsonInclude.Include.NON_NULL)
  private String rateDate;
  @Builder.Default
  private Collection<IsoCurrencyPair> pairs = new ArrayList<>();

  @JsonIgnore
  private IsoCurrencyPair tradePf;
  @JsonIgnore
  private IsoCurrencyPair tradeCash;
  @JsonIgnore
  private IsoCurrencyPair tradeBase;

  @JsonIgnore
  public FxRequest add(IsoCurrencyPair isoCurrencyPair) {
    if (isoCurrencyPair == null) {
      return this;
    }

    if (!pairs.contains(isoCurrencyPair)) {
      pairs.add(isoCurrencyPair);
    }

    return this;
  }

  public void setTradePf(IsoCurrencyPair tradePf) {
    this.tradePf = tradePf;
    add(tradePf);
  }

  public void setTradeBase(IsoCurrencyPair tradeBase) {
    this.tradeBase = tradeBase;
    add(tradeBase);
  }

  public void setTradeCash(IsoCurrencyPair tradeCash) {
    this.tradeCash = tradeCash;
    add(tradeCash);
  }
}

File: jar-common/src/main/kotlin/com/beancounter/common/contracts/FxRequest.kt
package com.beancounter.common.contracts

import com.beancounter.common.model.IsoCurrencyPair
import com.beancounter.common.utils.DateUtils
import com.fasterxml.jackson.annotation.JsonIgnore

/**
 * Represents a foreign exchange request for obtaining currency conversion rates for specified currency pairs on a given date.
 * This class stores a collection of unique currency pairs for which rates are requested and provides functionality to add currency pairs
 * related to various financial transactions like portfolio, cash management, and base currency operations.
 *
 * The currency pairs are stored in a way that ensures there are no duplicates, preserving the integrity of the request.
 *
 * Usage:
 * - `add(isoCurrencyPair: IsoCurrencyPair?)`: Adds a new currency pair to the request if it isn't already present.
 * - `addTradePf(tradePf: IsoCurrencyPair?)`: Adds a currency pair related to portfolio transactions and ensures it's included in the rate request.
 * - `addTradeBase(tradeBase: IsoCurrencyPair?)`: Adds a currency pair used as the base in transactions.
 * - `addTradeCash(tradeCash: IsoCurrencyPair?)`: Adds a currency pair related to cash transactions.
 *
 * @property rateDate The date for which the exchange rates are requested, defaulting to today.
 * @property pairs The set of unique currency pairs for which rates are needed.
 * @property tradePf Optional portfolio-related currency pair.
 * @property tradeCash Optional cash-related currency pair.
 * @property tradeBase Optional base currency pair.
 */
data class FxRequest(
    val rateDate: String = DateUtils.TODAY,
    val pairs: MutableSet<IsoCurrencyPair> = mutableSetOf(),
) {
    @JsonIgnore
    var tradePf: IsoCurrencyPair? = null

    @JsonIgnore
    var tradeCash: IsoCurrencyPair? = null

    @JsonIgnore
    var tradeBase: IsoCurrencyPair? = null

    @JsonIgnore
    fun add(isoCurrencyPair: IsoCurrencyPair?): FxRequest {
        if (isoCurrencyPair != null) {
            pairs.add(isoCurrencyPair)
        }
        return this
    }

    fun addTradePf(tradePf: IsoCurrencyPair?) {
        this.tradePf = tradePf
        add(tradePf)
    }

    fun addTradeBase(tradeBase: IsoCurrencyPair?) {
        this.tradeBase = tradeBase
        add(tradeBase)
    }

    fun addTradeCash(tradeCash: IsoCurrencyPair?) {
        this.tradeCash = tradeCash
        add(tradeCash)
    }
}
